### PR TITLE
COR-425-Fix_token_renewal

### PR DESCRIPTION
**Pull request by nilueps**
This probably isn't the most elegant solution, but I left the browser window open for over 30 minutes and I wasn't automatically logged out as was the case previously.

Edit: the refresh gets triggered, but I'm still getting an unauthorized response

### CODE REVIEW

**Comment by nilueps** [comment]
Don't merge this, I just realized that my fix isn't working for POST requests somehow

**Review by internetti — COMMENTED** [review]
I think it would be better to have a look at the TokenRequest.isTokenFresh function, maybe that one is not working correctly. we could write a test for it and see if it behaves like we expect.

since we actually have tokens with expiration dates, I think we should use those expiration dates instead of a fix interval

File: frontend/packages/core-auth/src/components/AuthWrapper.tsx [hunk 1]
```diff
@@ -98,17 +96,20 @@ const AuthWrapper: React.FC<AuthWrapperProps> = ({
...
           setTokenResponse(resp);
         })
-        .catch((err) => {
+        .catch(() => {
           setTokenResponse(undefined);
         });
     }
-  }, [tokenResponse?.shouldRefresh(), discovery]);
+  };
+  let refreshTokenIntervalId: NodeJS.Timeout | null = null;
```
Comment (neb42) [inline frontend/packages/core-auth/src/components/AuthWrapper.tsx:104]: I think you probably want this to be a `ref` so that is doesn't get set to null on each render.

```
const refreshTokenIntervalId = React.useRef<NodeJS.Timeout | null>(null);
...
if (refreshTokenIntervalId.current != null) clearInterval(refreshTokenIntervalId.current);
refreshTokenIntervalId.current = setInterval(()=>refreshToken(), 5000);
```

Comment (nilueps) [reply frontend/packages/core-auth/src/components/AuthWrapper.tsx:104]: That's neat, I didn't know about `useReft`, thanks

File: frontend/packages/core-auth/src/components/AuthWrapper.tsx [hunk 2]
```diff
@@ -83,29 +80,29 @@ const AuthWrapper: React.FC<AuthWrapperProps> = ({
...
+        break;
+      case 'id_token':
+        token = tokenResponse?.idToken ?? '';
+        break;
+      default:
+        token = '';
     }
-  }, [tokenResponse?.shouldRefresh(), discovery]);
+    onTokenChange(token);
+  }, [JSON.stringify(tokenResponse)]);
```
Comment (nilueps) [inline frontend/packages/core-auth/src/components/AuthWrapper.tsx:101]: The functionality is nearly there, but for some reason that I don't understand, this `useEffect` hook is not getting triggered when the timeout triggers the token refresh.

The `setTokenResponse` call on line 160 doesn't cause a re-render; which is what I would [expect](https://stackoverflow.com/questions/46240647/react-how-to-force-a-function-component-to-render/53837442#53837442)...  I must be overlooking something

File: frontend/packages/core-api-client/src/lib/BaseRESTClient.ts [hunk 3]
```diff
@@ -1,31 +1,40 @@
...
-      if (!token) {
-        return value;
-      }
-      return {
-        ...result,
-        headers: {
-          ...result.headers,
+    console.log('Setting token to:', token);
+
+    if (this.lastInterceptorId != null)
```
Comment (neb42) [inline frontend/packages/core-api-client/src/lib/BaseRESTClient.ts:20]: I think rather than ejecting the interceptor and using a new one, we could do a slight refactor. 

* Move the interceptor logic into the constructor
* Rather than taking `token` as a function argument, it's a method on the class `this.getToken()`. This will be called on each intercept.
* Add another class method `setToken` that will just store it as a property.
* When using `AuthWrapper` use `setToken` instead of `setAuth`

Comment (neb42) [reply frontend/packages/core-api-client/src/lib/BaseRESTClient.ts:20]: This package also seems useful, but I'm not sure it fits in with how we've structured auth. Potentially something for a refactor later down the line.
https://www.npmjs.com/package/axios-auth-refresh

**Comment by neb42** [comment]
The refresh logic should probably handle a response returning a `401`, refreshing the token, and trying again.

File: frontend/packages/core-auth/src/components/AuthWrapper.tsx (hunk 2 above)
Comment (neb42) [reply frontend/packages/core-auth/src/components/AuthWrapper.tsx:101]: `[JSON.stringify(tokenResponse)]` I don't think this will stringify that well. When using `JSON.stringify` it should be something json serialisable.

I'd replace this with `[tokenResponse?.accessToken, tokenResponse?.idToken]`

File: frontend/packages/core-auth/src/components/AuthWrapper.tsx [hunk 4]
```diff
@@ -42,27 +51,15 @@ const AuthWrapper: React.FC<AuthWrapperProps> = ({
...
-    switch (injectToken) {
-      case 'access_token':
-        return tokenResponse?.accessToken ?? '';
-      case 'id_token':
-        return tokenResponse?.idToken ?? '';
-      default:
-        return '';
-    }
-  })();
-
```
Comment (neb42) [inline frontend/packages/core-auth/src/components/AuthWrapper.tsx:55]: Why did you remove this block?

File: frontend/packages/core-auth/src/components/AuthWrapper.tsx [hunk 5]
```diff
@@ -114,16 +111,57 @@ const AuthWrapper: React.FC<AuthWrapperProps> = ({
...
+      if (delay <= 0) {
+        refreshToken();
+      } else {
+        refreshTokenTimeout.current = window.setTimeout(
+          () => refreshToken(),
+          delay,
+        );
+      }
+    }
+  }, [JSON.stringify(tokenResponse), discovery]);
```
Comment (neb42) [inline frontend/packages/core-auth/src/components/AuthWrapper.tsx:137]: Same here. I think `discovery` is ok, but you should check against the bits of `tokenResponse` you're interested in.

File: frontend/packages/core-auth/src/components/AuthWrapper.tsx [hunk 6]
```diff
@@ -114,16 +111,57 @@ const AuthWrapper: React.FC<AuthWrapperProps> = ({
     }
   }, [tokenResponse, isLoggedIn]);
 
+  // Schedule a token refresh before it expires (or use an existing schedule)
+  const refreshTokenTimeout = useRef<number | null>(null);
   useEffect(() => {
-    onTokenChange(token);
-  }, [token]);
+    if (
+      tokenResponse != null &&
```
Comment (neb42) [inline frontend/packages/core-auth/src/components/AuthWrapper.tsx:118]: All of these should be `!==` unless there's a good reason

File: frontend/packages/core-auth/src/components/AuthWrapper.tsx [hunk 7]
```diff
@@ -114,16 +111,57 @@ const AuthWrapper: React.FC<AuthWrapperProps> = ({
...
+    }
+  }, [JSON.stringify(tokenResponse), discovery]);
 
   const handleLogin = useCallback(() => {
     promptAsync().catch((err) => {
       handleLoginErr(err);
     });
   }, [discovery, request, promptAsync]);
 
+  async function refreshToken() {
```
Comment (neb42) [inline frontend/packages/core-auth/src/components/AuthWrapper.tsx:145]: This is defined after it is used. I think this is a linting error and generally makes the code harder to read.

File: frontend/packages/core-auth/src/components/AuthWrapper.tsx [hunk 8]
```diff
@@ -114,16 +111,57 @@ const AuthWrapper: React.FC<AuthWrapperProps> = ({
...
+      console.log({ delay });
+
+      if (delay <= 0) {
+        refreshToken();
+      } else {
+        refreshTokenTimeout.current = window.setTimeout(
+          () => refreshToken(),
+          delay,
+        );
+      }
```
Comment (neb42) [inline frontend/packages/core-auth/src/components/AuthWrapper.tsx:135]: I feel this is going to end up with some failed requests if you make a request while refreshing. It may be better to refresh more frequently while the access token is still valid. Eg. do this on a `setInterval` at 1/3 of the expiration time and always refresh.

File: frontend/packages/core-auth/src/components/AuthWrapper.tsx (hunk 4 above)
Comment (nilueps) [reply frontend/packages/core-auth/src/components/AuthWrapper.tsx:55]: I didn't, it just moved to line 89

File: frontend/packages/core-auth/src/components/AuthWrapper.tsx (hunk 6 above)
Comment (nilueps) [reply frontend/packages/core-auth/src/components/AuthWrapper.tsx:118]: Switch to truthy checks for object and `== / !=` for primitives. https://basarat.gitbook.io/typescript/styleguide#null-vs.-undefined

File: frontend/packages/core-api-client/src/lib/BaseRESTClient.ts [hunk 9]
```diff
@@ -1,38 +1,49 @@
...
 
 import { RequestOptions, Response } from './types';
 import { clientResponse } from './utils/responses';
 
 export class BaseRESTClient {
   protected readonly axiosInstance: AxiosInstance;
 
+  private token = '';
+
+  lastInterceptorId: number | undefined;
```
Comment (neb42) [inline frontend/packages/core-api-client/src/lib/BaseRESTClient.ts:11]: This isn't needed

File: frontend/packages/core-auth/src/types/response.ts [hunk 10]
```diff
@@ -23,15 +28,18 @@ export class TokenResponse implements TokenResponseConfig {
   public constructor(response: TokenResponseConfig) {
     this.accessToken = response.accessToken;
     this.tokenType = response.tokenType ?? TokenType.Bearer;
-    this.expiresIn = response.expiresIn;
+    this.expiresIn = 10; // response.expiresIn;
```
Comment (neb42) [inline frontend/packages/core-auth/src/types/response.ts:31]: I don't think this should have been committed?

File: frontend/packages/core-auth/src/types/response.ts [hunk 11]
```diff
@@ -83,15 +92,30 @@ export class TokenResponse implements TokenResponseConfig {
...
+      response = await request.performAsync(discovery);
+    } catch (e) {
+      throw new Error(`Error encountered while performing token refresh: ${e}`);
+    }
+
     response.refreshToken = response.refreshToken ?? this.refreshToken;
     const json = response.getRequestConfig();
     this.applyResponseConfig(json);
+
     return this;
```
Comment (neb42) [inline frontend/packages/core-auth/src/types/response.ts:119]: Why not this? The bit after the try/catch seems like it will never run on error and if it did then `response` could be `undefined` and we would error on `response.refreshToken = response.refreshToken ?? this.refreshToken;`. Putting it all in the try prevents this and also allows the response to be a `const` (I typically try to avoid using `let` unless a variable has a strong case for being reassigned).
```
try {
  const response = await request.performAsync(discovery);
  response.refreshToken = response.refreshToken ?? this.refreshToken;
  const json = response.getRequestConfig();
  this.applyResponseConfig(json);
  return this;
} catch (e) {
  throw new Error(`Error encountered while performing token refresh: ${e}`);
}
```

File: frontend/packages/core-auth/src/types/response.ts [hunk 12]
```diff
@@ -83,15 +92,30 @@ export class TokenResponse implements TokenResponseConfig {
     return !(TokenResponse.isTokenFresh(this) || !this.refreshToken);
   }
 
+  getExpiryMs(): number {
+    const tenPercentMargin = 0.9;
```
Comment (neb42) [inline frontend/packages/core-auth/src/types/response.ts:96]: Not sure I understand what the `tenPercentMargin` is for. I'd expect this method to give `expiresIn` in milliseconds. Returning 90% of the value isn't clear.

**Comment by neb42** [comment]
I think something like this would clean the refresh `useEffect` up. I don't think you really need to be checking the expiration time. We can just blindly refresh it.

```
  React.useEffect(() => {
    const refreshToken = async () => {
      if (!discovery) return;
      if (!tokenResponse) return;
      if (!tokenResponse.expiresIn || tokenResponse.expiresIn <= 0) return;

      const refreshConfig = {
        clientId,
        scopes,
        extraParams: {},
      };

      try {
        const resp = await tokenResponse.refreshAsync(refreshConfig, discovery);
        setTokenResponse(resp);
      } catch {
        setTokenResponse(undefined);
      }
    };

    let interval: number | undefined;
    if (discovery && tokenResponse?.expiresIn) {
      interval = setInterval(refreshToken, tokenResponse.expiresIn / 3);
    }

    return () => {
      if (interval) clearInterval(interval);
    };
  }, [JSON.stringify(discovery), tokenResponse?.expiresIn]);
```

File: frontend/apps/core-app/src/components/AuthWrapper/useTokenResponse.ts [hunk 13]
```diff
@@ -69,29 +69,44 @@ export const useTokenResponse = (): [
...
+
+    if (refreshTokenInterval.current)
+      window.clearInterval(refreshTokenInterval.current);
+
+    if (
+      tokenResponse &&
+      tokenResponse.expiresIn &&
+      tokenResponse.expiresIn > 0
+    ) {
+      refreshTokenInterval.current = window.setInterval(refreshToken, 1000);
```
Comment (internetti) [inline frontend/apps/core-app/src/components/AuthWrapper/useTokenResponse.ts:102]: so
